napari/components/_tests/test_layer_slicer.py

```diff
@@ -0,0 +1,262 @@
+    yield layer_slicer
+    layer_slicer.shutdown()
+
+
+def test_slice_layers_async_with_one_async_layer(layer_slicer):
+    layer = FakeAsyncLayer()
+
+    future = layer_slicer.slice_layers_async(layers=[layer], dims=Dims())
+
+    assert future.result()[layer].id == 1
+
+
+def test_slice_layers_async_with_multiple_async_layer(layer_slicer):
+    layer1 = FakeAsyncLayer()
+    layer2 = FakeAsyncLayer()
+
+    future = layer_slicer.slice_layers_async(
+        layers=[layer1, layer2], dims=Dims()
+    )
+
+    assert future.result()[layer1].id == 1
+    assert future.result()[layer2].id == 1
+
+
+def test_slice_layers_async_emits_ready_event_when_done(layer_slicer):
+    layer = FakeAsyncLayer()
+    event_result = None
+
+    def on_done(event):
+        nonlocal event_result
+        event_result = event.value
+
+    layer_slicer.events.ready.connect(on_done)
+
+    future = layer_slicer.slice_layers_async(layers=[layer], dims=Dims())
+    actual_result = future.result()
+
+    assert actual_result is event_result
+
+
+def test_slice_layers_async_with_one_sync_layer(layer_slicer):
+    layer = FakeSyncLayer()
+    assert layer.slice_count == 0
+
+    future = layer_slicer.slice_layers_async(layers=[layer], dims=Dims())
+
+    assert layer.slice_count == 1
+    assert not future.result()
+
+
+def test_slice_layers_async_with_multiple_sync_layer(layer_slicer):
+    layer1 = FakeSyncLayer()
+    layer2 = FakeSyncLayer()
+    assert layer1.slice_count == 0
+    assert layer2.slice_count == 0
+
+    future = layer_slicer.slice_layers_async(
+        layers=[layer1, layer2], dims=Dims()
+    )
+
+    assert layer1.slice_count == 1
+    assert layer2.slice_count == 1
+    assert not future.result()
+
+
+def test_slice_layers_async_with_mixed_layers(layer_slicer):
+    layer1 = FakeAsyncLayer()
+    layer2 = FakeSyncLayer()
+    assert layer1.slice_count == 0
+    assert layer2.slice_count == 0
+
+    future = layer_slicer.slice_layers_async(
+        layers=[layer1, layer2], dims=Dims()
+    )
+
+    assert layer1.slice_count == 1
+    assert layer2.slice_count == 1
+    assert future.result()[layer1].id == 1
+    assert future.result().get(layer2, 999) == 999
+
+
+def test_slice_layers_async_lock_blocking(layer_slicer):
+    dims = Dims()
+    layer = FakeAsyncLayer()
+
+    assert layer.slice_count == 0
+    with layer.lock:
+        blocked = layer_slicer.slice_layers_async(layers=[layer], dims=dims)
+        assert not blocked.done()
+
+    assert layer.slice_count == 1
+    assert blocked.result()[layer].id == 1
+
+
+def test_slice_layers_async_multiple_calls_cancels_pending(layer_slicer):
+    dims = Dims()
+    layer = FakeAsyncLayer()
+
+    with layer.lock:
+        blocked = layer_slicer.slice_layers_async(layers=[layer], dims=dims)
+        pending = layer_slicer.slice_layers_async(layers=[layer], dims=dims)
+        layer_slicer.slice_layers_async(layers=[layer], dims=dims)
+        assert not blocked.done()
```

Comment:
As soon as a task is submitted, it goes into the `running` state. However, with the `layer.lock` on, the task cannot complete. Therefore, the `blocked` object here would be more accurately called `running_but_blocked`. 

Then, when the second task is created (line 170), we recognize that this task is already ini `task_to_layers` and we call `task.cancel()`. However, because the task is `running`, it can't be canceled ([see docs](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.running)). 

Now despite the task not being canceled, the code continues in `slice_layers_async` which then overrides the first task in `task_to_layers`. We now effectively have lost a handle on the first task which is still blocked.

Reply:
Actually, I just realized, this was caused by my implementation of the `task_to_layer` lock. After removing it, I get 

```
>       for task_layers, task in self._layers_to_task.items():
E       RuntimeError: dictionary changed size during iteration
```

Which indicates its need. I'll think a bit more on how that lock can be properly implemented.

Reply:
I see. But how much do we care about that handle to the first task? I think it's only there to handle the cancellation logic.

`_on_slice_done` will still be called when it actually finishes, which should be after we release the layer's lock. We may choose to not emit the event if the task cannot be found (as I think you're doing now), though that tended to create less desirable behavior in practice (because having some updates sometimes is nice).

Reply:
You'll probably want a utility function to find the existing task to cancel - if it can't find it, then you can return `None`. That way you can iterate over the dict with the lock, but cancel without it.

In general, I'd recommend only holding onto locks as briefly as you can and make sense for a block of code.